### DISCUSSION TITLE
tutorials/lift-off-part1/09-codegen

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#57] [course feedback]
# tutorials/lift-off-part1/09-codegen

https://www.apollographql.com/tutorials/lift-off-part1/09-codegen

**Comment by jottaxwds** [comment]
Got an error removing `ignoreNoDocuments`:

```
> catstronauts-client@1.0.0 generate
> graphql-codegen

✔ Parse Configuration
⚠ Generate outputs
  ❯ Generate to ./src/__generated__/
    ✔ Load GraphQL schemas
    ✖
      Unable to find any GraphQL type definitions for the following pointers:
      - src/**/*.tsx
    ◼ Generate
```

**Reply by lizhennessy** [reply]
Hi there @jottaxwds! This error indicates that there weren't yet any GraphQL documents found in our frontend code - at this stage, that's just fine! We removed the `ignoreNoDocuments` flag so that we could see the error output when/if the Code Generator is unable to find anything to scan. In the next lesson, we'll define our first GraphQL operation that the Code Generator can work its magic on. 

I'll make an update to this course to clarify the error we see and why it occurs at this stage. If you're still running into trouble with this command, please don't hesitate to reach out! I'm happy to take a look.

**Comment by Adit0507** [comment]
Hey there, I just finished this section but I haven't gotten a tick yet.

Will it affect my Section completion ?

**Reply by mabuyo** [reply]
Hey there @Adit0507! There are no tasks or assessments in the codegen lesson, so that's why you're only seeing the "-" symbol in the lesson navigation. It won't affect your course completion status, you're good to go!

**Reply by Adit0507** [reply]
Ok thank you 😊

**Comment by amt-dwemoh-kesseh** [comment]
great

**Comment by GermanParlatto** [comment]
GraphQL is awesome !

**Comment by atgreen10** [comment]
Hey there, working through this module, and after running <npm install -D @graphql-codegen/cli @graphql-codegen/client-preset> and adding the import statement at the top of the codegen.ts file, I am receiving an error stating that the module cannot be found still. I've already checked the package.json file as well and can see the module has been installed under the devDependencies but the module continues to not be found. Any ideas on this one?

**Reply by GermanParlatto** [reply]
Hey, 

Your import should be like : 
`import { CodegenConfig } from "@graphql-codegen/cli";`

You can try to restart the TypeScript compiler sometimes it's not getting the updated properly. If you are using VSCode, access by 'shift + command + P' then type typescritp... and look for "TypeScript: Restart TS Server"
 
Cheers!

**Reply by TerrenceMM2** [reply]
Hey, @atgreen10. I got the same error when trying to install the client dependencies. Try installing each package individually.

**Reply by Navedphysicist** [reply]
I am also getting the same error

I have installed all the packages as per required but I am getting an error as : Cannot find module '@graphql-codegen/cli' or its corresponding type declarations.ts
I don't know why it is showing me this error . I have restarted the typescript server also but still the error persist
I can't apply npm run generate it is showing
npm ERR! Missing script: "generate"
But I have scripts as generate

**Comment by Ashuku001** [comment]
Hi there does this configuration generate types for mutation requests

**Reply by mabuyo** [reply]
Hey @Ashuku001 - yes it will! In the config, we've set it to generate types for any files inside `src` that end in `tsx`. 

```js
documents: ["src/**/*.tsx"]
```

Note that we don't get into mutations in this course, but this config will handle mutations when we get to it in a later course.

**Comment by MadFunx** [comment]
Hello, any idea on why Codegen doesn't retrieve Arguments description from a NestJS Apollo server?
In the .gql, there are comments above each argument of a query/mutation, but in the generated file they disappear, but all other descriptions are well retrieved.
Thank you by advance !

**Reply by lizhennessy** [reply]
Hey there @MadFunx ! Let's troubleshoot. Could you let me know what operating system you're using, and your Node version? (And if you have your code pushed to a remote repository, that would be great too - I'd love to take a look.)

**Comment by serabakpak** [comment]
Hello, you mentioned that you'll be covering fragments in an upcoming course. Could you point me in its direction? Thank you.

**Reply by mabuyo** [reply]
Hey @serabakpak - check out the last lesson of the [Side Quest: Intermediate Schema Design](https://www.apollographql.com/tutorials/side-quest-intermediate-schema-design/05-query-fragments), or the [Fragments section on graphql.com](https://graphql.com/learn/interfaces-and-unions/#fragments)!

**Reply by serabakpak** [reply]
Thank you for the quick response!

**Comment by Navedphysicist** [comment]
Hii there , 
I have installed all the packages as per required but I am getting an error as : Cannot find module '@graphql-codegen/cli' or its corresponding type declarations.ts
I don't know why it is showing me this error . I have restarted the typescript server also but still the error persist
I can't apply npm run generate it is showing 
npm ERR! Missing script: "generate"
But I have scripts as generate

**Reply by lizhennessy** [reply]
Hi @Navedphysicist! It sounds like the installation of the `@graphql-codegen/cli` package might have failed. Could you make sure that you've navigated to the `client` folder before running this command?

```
npm install -D @graphql-codegen/cli @graphql-codegen/client-preset
```

I noticed that for some learners it works better to run each command separately, so we can also try this!

```
npm install -D @graphql-codegen/cli
```

then:

```
npm install -D @graphql-codegen/client-preset
```

Let me know if you're still running into errors afterwards! I'm happy to troubleshoot further.

**Comment by Dark417** [comment]
Love it but '2m' is definitely a lie ..

**Reply by lizhennessy** [reply]
Thanks for calling this out! We've made an update to our time estimate.

**Comment by belloshehu** [comment]
Is codegen specific to Apollo or GraphQL generally? It is awesome though!

**Reply by lizhennessy** [reply]
Hi there @belloshehu ! Great question; lots of different languages and frameworks take advantage of some code generation process to help ease the burden on the developer - particularly when there's lots of manual work involved! In this course, the codegen that we're using _is_ specific to GraphQL; it's a tool that was developed by [The Guild](https://the-guild.dev/graphql/codegen/docs/getting-started) and it definitely makes generating the types from our GraphQL schema super easy. I'm glad that you enjoy it!

**Reply by belloshehu** [reply]
Thanks so much for the quick response! I hope lifting the burden does not deprive new developers from learning what is happening under the hood.

**Comment by SKumar1020** [comment]
when I run "npm run generate" with " ignoreNoDocuments: true" in "codege.ts", I get the following error 
--------------------------------------
✔ Parse Configuration                                        
⚠ Generate outputs                                           
  ✖ Unable to find preset matching client                    
                                                             
          Unable …                        
--------------------------------------

**Reply by lizhennessy** [reply]
Hi there @SKumar1020 ! I am happy to take a look and help troubleshoot. Do you have your code pushed up to a repository that I can take a look at?

**Reply by SKumar1020** [reply]
Thank you Liz. The error was related to my environment and I switched to a clean environment and got the issue resolved.

**Comment by antidiestro** [comment]
What is the next lesson? I click on Next and just get redirected to the first lesson in the Liftoff course.

**Reply by mabuyo** [reply]
Hey there @antidiestro - I can't seem to reproduce your issue, it should navigate you to the next lesson in the course: https://www.apollographql.com/tutorials/lift-off-part1/09-defining-a-query You can also use the lesson navigation in the sidebar if the Next button isn't working.

Curious to hear what browser are you using? Are you logged in with your Apollo account?

**Comment by dorakadri** [comment]
hello im stuck to days in this kind of error when i generate subscription i get the args the type subscription { ...} but not the actual type can anyone help please

**Reply by lizhennessy** [reply]
Hi there @dorakadri ! Happy to help troubleshoot. Do you have a link to your repo that I could check out?

**Comment by SOMONSOUM** [comment]
Hello, is there anyone who could help me with this error ?
I got failed to load schema from my graphql api endpoint like this
 Generate outputs                                                                                                                                                                                                        
  ❯ Generate to ./src/graphql/tou/types/                                                                                                                                                                                  
    ✖                                                                                                                                                                                                                     
            Failed to load schema from https:/tou-graphql.uat.puthi.online/graphql:                                                                                                                                        
                                                                                                                                                                                                                          
            Context creation failed: Cannot read properties of undefined (reading 'includes')                                                                                                                             
            GraphQLError: Context creation…                                                                                                                                                                               
    ◼ Load GraphQL documents                                                                                                                                                                                              
    ◼ Generate                                                                                                                                                                                                            
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

and this what I do in codegen.ts
```
import { CodegenConfig } from "@graphql-codegen/cli";
const token =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MiwidXNlcm5hbWUiOiJzb2suZGFyYUBtb2MuZ292LmtoIiwicGFzc3dvcmQiOiIkMmEkMTIkNlA5aGdhRDBTaDA5bnpYazlrUGZRT1I5THBpa0tzS1BNdU1uT3NvWUE1cGZmNTFURWE3VlciLCJmdWxsbmFtZSI6IlNvayBEYXJhIiwiY3JlYXRlZF9hdCI6IjIwMjQtMDItMTNUMDc6MDA6NTguMDAwWiIsInVwZGF0ZWRfYXQiOiIyMDI0LTAyLTEzVDA3OjAwOjU4LjAwMFoiLCJlbWFpbCI6IiIsInBob25lX251bWJlciI6IiIsInByb2ZpbGUiOm51bGwsImdlbmRlciI6bnVsbCwibmF0aW9uYWxpdHkiOm51bGwsImRvYiI6bnVsbCwiY29udGFjdF9jaXR5X29yX3Byb3ZpbmNlIjpudWxsLCJjb250YWN0X2Rpc3RyaWN0IjpudWxsLCJjb250YWN0X2NvbW11bmUiOm51bGwsImNvbnRhY3RfdmlsbGFnZV9vcl9ncm91cCI6bnVsbCwiaXNfZGVsZXRlIjowLCJpc192ZXJpZnkiOjEsInN0b3JhZ2VfcW91dGEiOjEwMCwiaWF0IjoxNzA4MTM1ODQzLCJleHAiOjE3MTA3Mjc4NDN9.td5xYZiQDfsNIHUg82agtf03v9oV2uzMMz92We9xwF4";

const config: CodegenConfig = {
  schema: [
    {
      "https:/tou-graphql.uat.puthi.online/graphql": {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    },
  ],
  documents: ["src/graphql/**/*.tou.{.gql, .graphql}"],
  generates: {
    "./src/graphql/tou/types/": {
      preset: "client",
      plugins: [
        "typescript",
        "typescript-operations",
        "typescript-react-apollo",
        "typescript-resolvers",
      ],
      presetConfig: {
        gqlTagName: "gql",
      },
      config: {
        withHooks: true,
      },
    },
  },
  ignoreNoDocuments: true,
};

export default config;
```

**Comment by NishantCoder108** [comment]
Codegen

**Comment by jimmyzzxhlh** [comment]
I might be missing something because I couldn't find any detail in the following lessons, but can someone share how to address this issue? 
`Variable 'documents' implicitly has type 'any[]' in some locations where its type cannot be determined.`

**Reply by lizhennessy** [reply]
Hi there @jimmyzzxhlh ! Happy to take a look, could you share the contents of your `codegen.ts` file with me, or provide a link to your repo?

**Comment by Weissberger** [comment]
The "next" lesson appears to be an invalid link, can you please share documentation on how to add client queries to the process? Also what If I'm not using apollo client but only codegen?

**Reply by lizhennessy** [reply]
Hi there @Weissberger ! Are you able to access the link directly? Here it is: https://www.apollographql.com/tutorials/lift-off-part1/09-defining-a-query . You can still use codegen in the frontend application without using Apollo Client explicitly; whatever tool you're using should have some guidance about integrating with GraphQL. Out of curiosity, is there a particular framework that you're looking to make this work with?

**Reply by Weissberger** [reply]
Hi @lizhennessy yes I can access the link, thanks for sharing. Currently we're not using a framework and doing this natively since we only have one mutation which doesn't need to be cached (yet we are currently looking into `apollo-client` for our subscription). We're using aws appsync for our graphql server, wondering if there's any guidance on how to integrate with that?

**Comment by cdwheatley** [comment]
I am getting an error on `gqlTagName` as follows.  Anyone know what this should be updated to?
```
Invalid Codegen Configuration!
  Please make sure that your codegen config file has defined plugins list for output "presetConfig".
  It should looks like that:
  schema:
  - my-schema.graphql
  generates:
  my-file.ts:
  - plugin1
  - plugin2
  - plugin3
◼ Generate outputs
```

**Reply by cdwheatley** [reply]
Well, nevermind, just figured out it was an incorrect schema location

**Comment by rubenesda** [comment]
It worked fine

**Comment by gustvoassuncao** [comment]
It worked fine for me, awesome tutorial, thanks!

**Comment by davinotdavid** [comment]
Hi! Thanks for sharing such a great tutorial, I am really enjoying so far!

As a heads up on the "Show code for codegen.ts" block portion – it seems like we are missing a comma on Line 13 to make it a valid object, right before `ignoreNoDocuments: true` in case anyone copy / pastes directly.

**Reply by lizhennessy** [reply]
Thank you for this great catch @davinotdavid ! Shipping that comma momentarily. :)

**Comment by zackmckennarunpod** [comment]
Hey! Im migrating from some other codegen code I found:
```
  schema: [
    {
      [graphqlUriWithApiKey]: {},
    },
  ],
  documents: "src/graphql/**/*.ts", // Path to your GraphQL queries/mutations
  generates: {
    "src/generated/graphql.tsx": {
      plugins: ["typescript", "typescript-operations", "typescript-react-apollo"],
    },
  },
```

I've updated it for the client preset:
```
  schema: [
    {
      [graphqlUriWithApiKey]: {},
    },
  ],
  documents: ["src/**/*.ts", "src/**/*.tsx"], // Path to your GraphQL queries/mutations
  generates: {
    "./src/__generated__/": {
      preset: "client",
      presetConfig: {
        gqlTagName: "gql",
      },
    },
  },
  ignoreNoDocuments: true,
```
but now when I run the generator, my basic types are not generated (export type User, export type Endpoint), where with the previous config they were. only the input types are generated. Any thoughts about why this might be?

**Reply by lizhennessy** [reply]
Hmmm @zackmckennarunpod interesting; so by the looks of it, your GraphQL server is running remotely at some address (`graphqlUriWithApiKey`), is that right? We do something similar in the [dedicated frontend version of this course](https://www.apollographql.com/tutorials/client-side-graphql-react/05-codegen) which might be a bit more helpful in this case. I'd be curious to see the output of your `__generated__` folder. Do you have the code pushed up to a public repo?

**Reply by zackmckennarunpod** [reply]
Hey! We don't have it in a public repo at this time, but can look into the `__generated__` file to look for something in particular. In regards to this issue, it turns out the types in codegen client preset were not expected to be used apparently: [github issue link](https://github.com/dotansimha/graphql-code-generator/issues/10167) or at least the generated gql types were removed entirely from gql codegen in the newest version. I got around it by running both the client preset and typescript plugin next to each other and running a script to add the generated types to my __generated__ file. Thanks for the link to the course, I will definitely check this out!

**Reply by zackmckennarunpod** [reply]
[comment about not using the client preset types](https://github.com/dotansimha/graphql-code-generator/issues/10167#issuecomment-2416574921)

**Comment by artslimedev** [comment]
May want to add for the people using typescript to include plugins that support it as mine was not generating the graphql.ts without it. 
`plugins: [
        'typescript',
        'typescript-operations',
        'typescript-react-apollo',
      ],`

**Reply by bbil** [reply]
+1 to that, I was a bit stumped for a minute.

**Comment by Arcanorum** [comment]
Why is this ugly "./src/\_\_generated\_\_" directory name needed?
Why not just "./src/gql" as shown in https://the-guild.dev/graphql/codegen/docs/getting-started ?

**Reply by lizhennessy** [reply]
hey @Arcanorum ! That's totally valid too. If you prefer `gql`, definitely do so! In our example, `__generated__` is intended to make it really really clear that our directory contains just that, generated code. (As opposed to GraphQL schemas, or operations)

**Comment by b0nec0de** [comment]
Hi, while trying to run 'generate' script I'm getting an error: `Failed to load schema from http://localhost:4000: AbortSignal.any is not a function`. Could you please shed some light on this issue?

**Reply by mabuyo** [reply]
Hi there @b0nec0de! Sorry to hear you're having issues. I can't find anything around that error message unfortunately. Could you please double check the following:

- installed the correct packages: `npm install -D @graphql-codegen/cli @graphql-codegen/client-preset @graphql-codegen/typescript @graphql-codegen/typescript-operations`
- `codegen.ts` contains the following:

```ts
import { CodegenConfig } from "@graphql-codegen/cli";

const config: CodegenConfig = {
  schema: "http://localhost:4000",
  documents: ["src/**/*.tsx"],
  generates: {
    "./src/__generated__/": {
      preset: 'client',
      presetConfig: {
        gqlTagName: "gql"
      }
    },
    "./src/__generated__/types.ts": {
      plugins: ["typescript", "typescript-operations"],
    },
  },
  ignoreNoDocuments: true,
};

export default config;
```

- The server is running on `localhost:4000`. 

Hopefully that helps!

**Reply by b0nec0de** [reply]
Hey, thanks for your suggestions! The real problem was in the node version though. I updated it to the latest and now it works. 
I walked around it before that by direct referencing the schema file, located in the `server/schema.ts`. I know, that is what you don't recommend to do, but this is almost exactly what I have on my previous project: `schema.ts` file was located in the code base on frontend and after each change of schema on backend we had to download the new schema file and replace the previous one with it (and fix conflicts). It was not a best developer experience :)

**Reply by mabuyo** [reply]
I'm glad updating node did the trick! Re: the schema file, yes in our case, we opted to generate types by connecting to the running server so we get the most up-to-date schema from there. 

Enjoy the rest of the course, let us know if you run into any other issues!

**Comment by sw-3** [comment]
I get the same error as the person above.  The codegen.ts file is correct, and i followed the instructions exactly with the installs by copy/pasting.
I restarted the server, and got the same error.   The output of npm run generate is:

> catstronauts-client@1.0.0 generate
> graphql-codegen

✔ Parse Configuration
⚠ Generate outputs
  ❯ Generate to ./src/__generated__/
    ✖
      Failed to load schema from http://localhost:4000:
      AbortSignal.any is not a function
      TypeError: AbortSignal.any is not a function
      at baseExecutor (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/executor-http/dist/index.cjs:333:32)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/url-loader/cjs/index.js:163:103
      at mapMaybePromise (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/utils/cjs/map-maybe-promise.js:25:12)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/url-loader/cjs/index.js:163:59
      at schemaFromExecutor (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/wrap/dist/index.cjs:2415:7)
      at UrlLoader.load (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/url-loader/cjs/index.js:206:65)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/load/cjs/load-typedefs/load-file.js:15:52
      at Array.map (<anonymous>)
      at loadFile (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/load/cjs/load-typedefs/load-file.js:13:43)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/load/cjs/load-typedefs/collect-sources.js:200:59
      GraphQL Code Generator supports:
      - ES Modules and CommonJS exports (export as default or named export "schema")
      - Introspection JSON File
      - URL of GraphQL endpoint
      - Multiple files with type definitions (glob expression)
      - String in config file
      Try to use one of above options and run codegen again.
    ◼ Load GraphQL documents
    ◼ Generate
  ❯ Generate to ./src/__generated__/types.ts
    ✖
      Failed to load schema from http://localhost:4000:
      AbortSignal.any is not a function
      TypeError: AbortSignal.any is not a function
      at baseExecutor (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/executor-http/dist/index.cjs:333:32)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/url-loader/cjs/index.js:163:103
      at mapMaybePromise (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/utils/cjs/map-maybe-promise.js:25:12)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/url-loader/cjs/index.js:163:59
      at schemaFromExecutor (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/wrap/dist/index.cjs:2415:7)
      at UrlLoader.load (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/url-loader/cjs/index.js:206:65)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/load/cjs/load-typedefs/load-file.js:15:52
      at Array.map (<anonymous>)
      at loadFile (/home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/load/cjs/load-typedefs/load-file.js:13:43)
      at /home/scottw3/devel/graphql/odyssey-lift-off-part1/client/node_modules/@graphql-tools/load/cjs/load-typedefs/collect-sources.js:200:59
      GraphQL Code Generator supports:
      - ES Modules and CommonJS exports (export as default or named export "schema")
      - Introspection JSON File
      - URL of GraphQL endpoint
      - Multiple files with type definitions (glob expression)
      - String in config file
      Try to use one of above options and run codegen again.
    ◼ Load GraphQL documents
    ◼ Generate

**Reply by sw-3** [reply]
note that the above interpreted the double-underline as markup ... the output really does have 2 underline characters before and after 'generated' in the directory name.

**Reply by mabuyo** [reply]
Hi there @sw-3 ! Could you confirm your node version? It looks like you'll need at least node `v20.3` for codegen.

**Reply by mabuyo** [reply]
@sw-3 following up, looks like a node upgrade to the latest version fixed the issue for the other person. Let us know if it works for you too!

**Reply by sw-3** [reply]
Hi!  Thanks so much for the info!!
(I have been fighting with the blockchain side of my project for the last 2 days, so did not get back to the graphql part yet.  Ha!)

I think the default node version for my OS version is 18.x, so that makes sense!  I can update it easy enough and will try it out when I get time.  cheers!

**Comment by paulnordlund** [comment]
Wondering if anyone is getting an error about the imports in both client __generated gql.ts and graphql.ts:  import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core'; ?  Any workarounds?  There seems to be some discussion about a codegen issue online but adding this to codegen.ts isn't working for me:  config: { useTypeImports: true },

**Reply by paulnordlund** [reply]
Discussion: https://github.com/dotansimha/graphql-typed-document-node/issues/152#issuecomment-1446270040

**Reply by paulnordlund** [reply]
Found a solution: in tsconfig.json, comment out "verbatimModuleSyntax": true

**Comment by myklt** [comment]
According to the [latest codegen docs](https://www.apollographql.com/docs/react/development-testing/graphql-codegen#a-note-about-the-client-preset), it is not recommended to use `client` preset. Should this tutorial be updated respectively? Happy to contribute.
